### PR TITLE
Machine Time Interrupts for Enclaves

### PR DESCRIPTION
**Pull request by ghost**
We now handle timer interrupts in the SM. An enclave can cause a DDOS attack on other enclaves/SM if it doesn't yield the CPU. Not delegating time interrupts to S-mode prevents this. SM now handles time interrupts. Before an enclave runs, the SM sets a timer interrupt with some time slice, the max QUANTA the enclave should run. If the enclave lives to this QUANTA without yielding the CPU, the enclave will receive a machine timer interrupt which isn't delegated to the enclave and the SM will have to step in.

### CODE REVIEW

File: machine/mtrap.c [hunk 1]
```diff
@@ -124,6 +124,83 @@ static void send_ipi_many(uintptr_t* pmask, int event)
   }
 }
 
+
```
Comment (dayeol) [inline machine/mtrap.c:127]: Please move this function to sm/trap.c as well if possible.

File: sm_c/enclave.c [hunk 2]
```diff
@@ -28,6 +29,18 @@ extern byte dev_public_key[PUBLIC_KEY_SIZE];
  *
  ****************************/
 
+/*
```
Comment (dayeol) [inline sm_c/enclave.c]: Remove comments

File: sm_c/enclave.c [hunk 3]
```diff
@@ -108,13 +128,21 @@ static inline void context_switch_to_host(uintptr_t* encl_regs,
...
   swap_prev_mepc(&enclaves[eid].threads[0], read_csr(mepc));
 
+  switch_vector_host(); 
+
+  uintptr_t interrupts = MIP_SSIP | MIP_STIP | MIP_SEIP;
+  write_csr(mideleg, interrupts);
+
   // enable timer interrupt
-  set_csr(mie, MIP_MTIP);
+  clear_csr(mie, MIP_MTIP);
```
Comment (dayeol) [inline sm_c/enclave.c]: Why are we disabling the timer interrupt here?

File: sm_c/enclave.h [hunk 4]
```diff
@@ -121,6 +121,11 @@ enclave_ret_code attest_enclave(uintptr_t report, uintptr_t data, uintptr_t size
 enclave_ret_code validate_and_hash_enclave(struct enclave* enclave);
 // TODO: These functions are supposed to be internal functions.
 void enclave_init_metadata();
+/*static inline void switch_vector_enclave();
```
Comment (dayeol) [inline sm_c/enclave.h]: Remove comments

File: sm_c/tests/CMakeLists.txt [hunk 5]
```diff
@@ -54,8 +55,11 @@ set_target_properties(test_enclave
         COMPILE_FLAGS -DTARGET_PLATFORM_HEADER=\\"..\/platform\/default\/default.h\\"
         LINK_FLAGS
           "-Wl,--wrap=printm \
-          -Wl,--wrap=poweroff \
-          -Wl,--wrap=disabled_hart_mask \
+  -Wl,--wrap=mtime   \
+          -Wl,--wrap=trap_vector \
+          -Wl,--wrap=trap_vector_enclave \
+	  -Wl,--wrap=poweroff \
```
Comment (dayeol) [inline sm_c/tests/CMakeLists.txt:61]: Please keep the indentation consistent (use space instead of tab)

File: sm_c/thread.c [hunk 6]
```diff
@@ -5,6 +5,36 @@
...
+}
+
+void swap_prev_mpp(struct thread_state* thread, uintptr_t* regs){
+  //Time interrupts can occur in either user mode or supervisor mode
+
+  int old_mpp = thread->prev_mpp;
+  int mstatus = read_csr(mstatus);
+  thread->prev_mpp = mstatus & 0x800;
+
+  if(old_mpp != 0) {
```
Comment (dayeol) [inline sm_c/thread.c]: I did not understand this part, why do you check (old_mpp != 0) ?
Why can't we just overwrite the MPP field of the mstatus with prev_mpp?
e.g.,  
new_mstatus = (mstatus & ~0x800) | thread->prev_mpp; 
csr_write(mstatus, new_mstatus);

File: sm_c/thread.c [hunk 7]
```diff
@@ -5,6 +5,36 @@
...
+     set_csr(mstatus, 0x800);
```
Comment (dayeol) [inline sm_c/thread.c]: This seems problematic for me, as it always sets MPP to a certain value (I believe S-mode) no matter what old_mpp previously was.

File: sm_c/thread.c [hunk 8]
```diff
@@ -71,6 +101,7 @@ void clean_state(struct thread_state* state){
     prev[i] = 0;
   }
 
+  state->prev_mpp = 1;
```
Comment (dayeol) [inline sm_c/thread.c]: See above comments

**Review by dayeol — CHANGES_REQUESTED** [review]

File: sm_c/enclave.c (hunk 3 above)
Comment (ghost) [reply sm_c/enclave.c]: Whoops! This was a remnant of my old code. Removing... We don't need to disable timer interrupts when in host context.

File: sm_c/enclave.h (hunk 4 above)
Comment (ghost) [reply sm_c/enclave.h]: Resolved.

File: machine/mtrap.c (hunk 1 above)
Comment (ghost) [reply machine/mtrap.c:127]: Just to confirm, but the function send_ipi_many?

File: sm_c/thread.c (hunk 6 above)
Comment (ghost) [reply sm_c/thread.c]: Yup, this would work. Yours is much more succint, so I changed it to this.

File: sm_c/thread.c (hunk 7 above)
Comment (ghost) [reply sm_c/thread.c]: You're right. I added logic to initialize the MPP. See swap_prev_mpp.

File: sm_c/thread.c (hunk 8 above)
Comment (ghost) [reply sm_c/thread.c]: See above.

**Pull request merged**